src/protocol/primitives/record-batch/record-batch.ts

```diff
@@ -153,6 +159,27 @@ export class RecordBatch implements Serializable {
   }
 
   public async serialize(buffer: WriteBuffer): Promise<void> {
+    if (this.baseTimestamp.value < 0n && this.baseTimestamp.value !== RecordBatch.NO_TIMESTAMP) {
+      throw new IllegalArgumentError(`Invalid message timestamp ${this.baseTimestamp.value}`);
+    }
+
+    if (
+      this.attributes.isTransactional === IsTransactional.Yes &&
+      this.producerId.value === RecordBatch.NO_PRODUCER_ID
+    ) {
+      throw new IllegalArgumentError('Cannot write transactional messages without a valid producer ID');
+    }
+
+    if (this.producerId.value !== RecordBatch.NO_PRODUCER_ID) {
+      if (this.producerEpoch.value === RecordBatch.NO_PRODUCER_EPOCH) {
```

Comment:
This is directly taken from the [Java library](https://github.com/a0x8o/kafka/blob/master/clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java#L372)